src/packet_analysis/protocol/ip/IP.cc

```diff
@@ -1,37 +1,275 @@
+
+	if ( packet->len < total_len + packet->hdr_size )
+		{
+		packet->Weird("truncated_IP", encapsulation);
+		return false;
+		}
+
+	// For both of these it is safe to pass ip_hdr because the presence
+	// is guaranteed for the functions that pass data to us.
+	uint16_t ip_hdr_len = ip_hdr->HdrLen();
+	if ( ip_hdr_len > total_len )
+		{
+		sessions->Weird("invalid_IP_header_size", ip_hdr, encapsulation);
+		return false;
+		}
+
+	if ( ip_hdr_len > len )
+		{
+		sessions->Weird("internally_truncated_header", ip_hdr, encapsulation);
+		return false;
+		}
+
+	if ( ip_hdr->IP4_Hdr() )
+		{
+		if ( ip_hdr_len < sizeof(struct ip) )
+			{
+			packet->Weird("IPv4_min_header_size");
+			return false;
+			}
+		}
+	else
+		{
+		if ( ip_hdr_len < sizeof(struct ip6_hdr) )
+			{
+			packet->Weird("IPv6_min_header_size");
+			return false;
+			}
+		}
+
+	// Ignore if packet matches packet filter.
+	detail::PacketFilter* packet_filter = sessions->GetPacketFilter(false);
+	if ( packet_filter && packet_filter->Match(ip_hdr, total_len, len) )
+		 return false;
+
+	if ( ! packet->l2_checksummed && ! detail::ignore_checksums && ip4 &&
+	     detail::in_cksum(reinterpret_cast<const uint8_t*>(ip4), ip_hdr_len) != 0xffff )
+		{
+		sessions->Weird("bad_IP_checksum", packet, encapsulation);
+		return false;
+		}
+
+	if ( discarder && discarder->NextPacket(ip_hdr, total_len, len) )
+		return false;
+
+	detail::FragReassembler* f = nullptr;
+
+	if ( ip_hdr->IsFragment() )
+		{
+		packet->dump_packet = true;	// always record fragments
+
+		if ( len < total_len )
+			{
+			sessions->Weird("incompletely_captured_fragment", ip_hdr, encapsulation);
+
+			// Don't try to reassemble, that's doomed.
+			// Discard all except the first fragment (which
+			// is useful in analyzing header-only traces)
+			if ( ip_hdr->FragOffset() != 0 )
+				return false;
+			}
+		else
+			{
+			f = detail::fragment_mgr->NextFragment(run_state::processing_start_time, ip_hdr, packet->data + packet->hdr_size);
+			IP_Hdr* ih = f->ReassembledPkt();
+			if ( ! ih )
+				// It didn't reassemble into anything yet.
+				return true;
+
+			ip4 = ih->IP4_Hdr();
+
+			// Switch the stored ip header over to the one from the
+			// fragmented packet.
+			delete ip_hdr;
+			ip_hdr = ih;
+
+			len = total_len = ip_hdr->TotalLen();
+			ip_hdr_len = ip_hdr->HdrLen();
+			packet->cap_len = total_len + packet->hdr_size;
+			packet->ip_hdr = ih;
+
+			if ( ip_hdr_len > total_len )
+				{
+				sessions->Weird("invalid_IP_header_size", ip_hdr, encapsulation);
+				return false;
+				}
+			}
+		}
+
+	detail::FragReassemblerTracker frt(f);
+
+	// We stop building the chain when seeing IPPROTO_ESP so if it's
+	// there, it's always the last.
+	if ( ip_hdr->LastHeader() == IPPROTO_ESP )
+		{
+		packet->dump_packet = true;
+		if ( esp_packet )
+			event_mgr.Enqueue(esp_packet, ip_hdr->ToPktHdrVal());
+
+		// Can't do more since upper-layer payloads are going to be encrypted.
+		return true;
+		}
+
+#ifdef ENABLE_MOBILE_IPV6
```

Comment:
Do you have an idea why we have this #eefine rather than always including the code?

Reply:
No clue. It's like that in the old code too.

Reply:
Think `ENABLE_MOBILE_IPV6` was just deemed experimental/hard-to-verify-it-works at the time it was introduced.  Maybe worth a separate task by now to look at either removing if now bitrotten/broken else graduate to the default config (if truly no other concern, like performance).

Reply:
Seperate ticket sounds good. If anybody wants to take closer look, we can attempt to graduate to default, otherwise I would just remove it after a quick ask to users if anybody happens to be using it (I bet: no).

Reply:
Opened as https://github.com/zeek/zeek/issues/1216, resolving this comment.